File: framework/languages/java/e2e/StoreFailure/Server/Consumer/src/main/java/systems/zlink/e2e/storefailure/consumer/ConsumerOptions.java
package systems.zlink.e2e.storefailure.consumer;

import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties("e2e")
public record ConsumerOptions(
    String rid,
    String httpEndpoint,
    String redisLocationEndpoint,
    String locationKeyPrefix,
    long redisCommandTimeoutMillis,
    long heartbeatMillis,
    long leaseTtlMillis,
    long pollingMillis,
    long storeFailureGraceMillis,
    String storeMode,
    String storeDelayControlFile,
    String logDir) {
    public ConsumerOptions {
        required(rid, "rid");
        required(httpEndpoint, "http-endpoint");
        required(redisLocationEndpoint, "redis-location-endpoint");
        required(locationKeyPrefix, "location-key-prefix");
        positive(redisCommandTimeoutMillis, "redis-command-timeout-millis");
        positive(heartbeatMillis, "heartbeat-millis");
        positive(leaseTtlMillis, "lease-ttl-millis");
        positive(pollingMillis, "polling-millis");
        positive(storeFailureGraceMillis, "store-failure-grace-millis");
        required(storeMode, "store-mode");
        required(logDir, "log-dir");
        storeDelayControlFile = storeDelayControlFile == null ? "" : storeDelayControlFile;
    }

    private static void required(String value, String name) {
        if (value == null || value.isBlank()) throw new IllegalArgumentException("e2e." + name + " is required");
    }

    private static void positive(long value, String name) {
        if (value <= 0) throw new IllegalArgumentException("e2e." + name + " must be positive");
    }
}

File: framework/languages/java/e2e-kotlin/DiscoveryRegistryHa/Server/Consumer/src/main/kotlin/systems/zlink/e2e/kotlin/discoveryregistryha/consumer/Configuration/ConsumerOptions.kt
package systems.zlink.e2e.kotlin.discoveryregistryha.consumer.Configuration

import systems.zlink.e2e.kotlin.discoveryregistryha.CliOptions

data class ConsumerOptions(
    val rid: String,
    val httpEndpoint: String,
    val redisLocationEndpoint: String,
    val locationKeyPrefix: String,
    val redisCommandTimeoutMillis: Long,
    val heartbeatMillis: Long,
    val leaseTtlMillis: Long,
    val pollingMillis: Long,
    val storeFailureGraceMillis: Long,
    val storeMode: String,
    val logDir: String,
) {
    companion object {
        fun parse(args: Array<String>): ConsumerOptions {
            val options = CliOptions.parse(args)
            return ConsumerOptions(
                rid = options.get("--consumer-rid", "consumer"),
                httpEndpoint = options.get("--http-endpoint"),
                redisLocationEndpoint = options.get("--redis-location-endpoint"),
                locationKeyPrefix = options.get("--location-key-prefix"),
                redisCommandTimeoutMillis = options.get("--redis-command-timeout-ms", "5000").toLong(),
                heartbeatMillis = options.get("--location-heartbeat-ms", "1000").toLong(),
                leaseTtlMillis = options.get("--location-lease-ttl-ms", "3000").toLong(),
                pollingMillis = options.get("--location-polling-ms", "500").toLong(),
                storeFailureGraceMillis = options.get("--location-store-failure-grace-ms", "6000").toLong(),
                storeMode = options.get("--location-store-mode", "watch"),
                logDir = options.get("--log-dir", "logs"),
            )
        }
    }
}
